### PR TITLE
Implement DuckDB connector

### PR DESCRIPTION
**Pull request by grieve54706**
This PR supports DuckDB as datasource. It provides Restful API to set initialization SQL and session SQL. And provide API to directly query DuckDB. Also, support using wire protocol to connect through Accio MDL.

## Configuration
`duckdb.connector.init-sql-path`: default `etc/duckdb/init.sql`
`duckdb.connector.session-sql-path`: default `etc/duckdb/session.sql`

### `GET` /v1/data-source/DuckDB/settings/init-sql
Get Init-SQL
```sql
INSTALL httpfs;
CREATE TABLE tpch.orders AS SELECT * FROM read_parquet('/path-to-file/orders.parquet');
CREATE TABLE tpch.customer AS SELECT * FROM read_parquet('/path-to-file/customer');
```

### `PATCH` /v1/data-source/DuckDB/settings/init-sql
Run and Update Init-SQL (PATCH)
```sql
CREATE TABLE tpch.part AS SELECT * FROM read_parquet('/path-to-file/part.parquet');
```
Using this API, user can modify their DuckDB environment in runtime. Any SQL will be executed in DuckDB and append to the original init-SQL file for restarting server next time.

### `PUT` /v1/data-source/DuckDB/settings/init-sql
Overwrite Init-SQL (PUT)
```sql
INSTALL httpfs;
CREATE TABLE tpch.orders AS SELECT * FROM read_parquet('/path-to-file/orders.parquet');
CREATE TABLE tpch.customer AS SELECT * FROM read_parquet('/path-to-file/customer');
```
Clean the current DuckDB environment and create a new DuckDB instance by the new init-SQL. The old init-SQL will be archived and replaced by the new one.

### `GET` /v1/data-source/DuckDB/settings/session-sql
Get session SQL
```sql
SET s3_secret = 'xxxxx';
SET s3_accesskey = 'xxxx';
```

### `PATCH` /v1/data-source/DuckDB/settings/session-sql
Update session properties SQL (PATCH)
```sql
SET s3_secret = 'xxxxx';
```
Something like update init-SQL. It should be applied to all exist connection in pool.

### `PUT` /v1/data-source/DuckDB/settings/session-sql
Overwrite session properties SQL (PUT)
```sql
SET s3_secret = 'xxxxx';
SET s3_accesskey = 'xxxx';
```
Something like overwrite init-SQL. It should clear all exist connection in pool.

### `POST` /v1/data-source/DuckDB/query
Query DuckDB directly
```sql
SELECT * FROM memory.tpch.orders LIMIT 2;
```
```json
{
  "columns": [
    {
      "name": "o_orderkey",
      "type": "int4"
    },
    {
      "name": "o_custkey",
      "type": "int4"
    },
    {
      "name": "o_orderstatus",
      "type": "varchar"
    },
    {
      "name": "o_totalprice",
      "type": "numeric"
    },
    {
      "name": "o_orderdate",
      "type": "date"
    },
    {
      "name": "o_orderpriority",
      "type": "varchar"
    },
    {
      "name": "o_clerk",
      "type": "varchar"
    },
    {
      "name": "o_shippriority",
      "type": "int4"
    },
    {
      "name": "o_comment",
      "type": "varchar"
    }
  ],
  "rows": [
    [
      1,
      370,
      "O",
      172799.49,
      "1996-01-02",
      "5-LOW",
      "Clerk#000000951",
      0,
      "nstructions sleep furiously among "
    ],
    [
      2,
      781,
      "O",
      38426.09,
      "1996-12-01",
      "1-URGENT",
      "Clerk#000000880",
      0,
      " foxes. pending accounts at the pending, silent asymptot"
    ]
  ]
}
```

### CODE REVIEW

File: accio-main/src/main/java/io/accio/main/web/DuckDBResource.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    public void setInitSQL(
+            String sql,
+            @Suspended AsyncResponse asyncResponse)
+    {
+        metadata.setInitSQL(sql);
+        metadata.reload();
+        java.nio.file.Path initSQLPath = metadata.getInitSQLPath();
+        FileUtil.archiveFile(initSQLPath);
+        FileUtil.createFile(initSQLPath, sql);
+        asyncResponse.resume(Response.ok().build());
```
Comment (goldmedal) [inline accio-main/src/main/java/io/accio/main/web/DuckDBResource.java:80]: Same as above.

File: accio-main/src/main/java/io/accio/main/web/DuckDBResource.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    @PATCH
+    @Path("/settings/init-sql")
+    public void appendToInitSQL(
+            String sql,
+            @Suspended AsyncResponse asyncResponse)
+    {
+        metadata.directDDL(sql);
+        metadata.setInitSQL(metadata.getInitSQL() + "\n" + sql);
+        FileUtil.appendToFile(metadata.getInitSQLPath(), sql);
+        asyncResponse.resume(Response.ok().build());
```
Comment (goldmedal) [inline accio-main/src/main/java/io/accio/main/web/DuckDBResource.java:92]: Same as above.

File: accio-main/src/main/java/io/accio/main/web/DuckDBResource.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    @GET
+    @Path("/settings/session-sql")
+    public String getSessionSQL()
+    {
+        return metadata.getSessionSQL();
```
Comment (goldmedal) [inline accio-main/src/main/java/io/accio/main/web/DuckDBResource.java:99]: Same as above.

File: accio-main/src/main/java/io/accio/main/web/DuckDBResource.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    public void setSessionSQL(
+            String sql,
+            @Suspended AsyncResponse asyncResponse)
+    {
+        metadata.setSessionSQL(sql);
+        metadata.reload();
+        java.nio.file.Path sessionSQLPath = metadata.getSessionSQLPath();
+        FileUtil.archiveFile(sessionSQLPath);
+        FileUtil.createFile(sessionSQLPath, sql);
+        asyncResponse.resume(Response.ok().build());
```
Comment (goldmedal) [inline accio-main/src/main/java/io/accio/main/web/DuckDBResource.java:113]: Same as above.

File: accio-main/src/main/java/io/accio/main/web/DuckDBResource.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    @PATCH
+    @Path("/settings/session-sql")
+    public void appendToSessionSQL(
+            String sql,
+            @Suspended AsyncResponse asyncResponse)
+    {
+        metadata.directDDL(sql);
+        metadata.setSessionSQL(metadata.getSessionSQL() + "\n" + sql);
+        FileUtil.appendToFile(metadata.getSessionSQLPath(), sql);
+        asyncResponse.resume(Response.ok().build());
```
Comment (goldmedal) [inline accio-main/src/main/java/io/accio/main/web/DuckDBResource.java:125]: Same as above.

File: accio-main/src/main/java/io/accio/main/web/dto/DuckDBQueryDto.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
+package io.accio.main.web.dto;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.List;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+
+public class DuckDBQueryDto
```
Comment (goldmedal) [inline accio-main/src/main/java/io/accio/main/web/dto/DuckDBQueryDto.java:10]: This dto class is similar with `PreviewOutputDto`. Consider to merge them.

File: accio-base/src/main/java/io/accio/base/config/ConfigManager.java [hunk 7]
```diff
@@ -101,14 +104,16 @@ public ConfigManager(
...
     {
         this.accioConfig = Optional.of(accioConfig);
         this.postgresConfig = Optional.of(postgresConfig);
         this.bigQueryConfig = Optional.of(bigQueryConfig);
         this.duckDBConfig = Optional.of(duckDBConfig);
         this.postgresWireProtocolConfig = Optional.of(postgresWireProtocolConfig);
         this.duckdbS3StyleStorageConfig = Optional.of(duckdbS3StyleStorageConfig);
+        this.duckDBConnectorConfig = Optional.of(duckDBConnectorConfig);
 
         initConfig(
```
Comment (goldmedal) [inline accio-base/src/main/java/io/accio/base/config/ConfigManager.java:118]: I guess you may missing to init `DuckDBConnectorConfig`.

File: accio-tests/src/test/java/io/accio/testing/RequireAccioServer.java [hunk 8]
```diff
@@ -83,6 +87,7 @@ public abstract class RequireAccioServer
     private static final JsonCodec<ConfigManager.ConfigEntry> CONFIG_ENTRY_JSON_CODEC = jsonCodec(ConfigManager.ConfigEntry.class);
     private static final JsonCodec<List<ConfigManager.ConfigEntry>> CONFIG_ENTRY_LIST_CODEC = listJsonCodec(ConfigManager.ConfigEntry.class);
     private static final JsonCodec<PreviewOutputDto> PREVIEW_OUTPUT_DTO_CODEC = jsonCodec(PreviewOutputDto.class);
+    private static final JsonCodec<DuckDBQueryDto> DUCK_DB_QUERY_DTO_JSON_CODEC = jsonCodec(DuckDBQueryDto.class);
```
Comment (goldmedal) [inline accio-tests/src/test/java/io/accio/testing/RequireAccioServer.java:90]:
```suggestion
    private static final JsonCodec<DuckDBQueryDto> DUCKDB_QUERY_DTO_JSON_CODEC = jsonCodec(DuckDBQueryDto.class);
```
We use `DUCKDB` mostly.

File: accio-main/src/main/java/io/accio/main/web/DuckDBResource.java [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    private final DuckDBMetadata metadata;
+
+    @Inject
+    public DuckDBResource(
+            DuckDBMetadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+    }
+
+    @POST
```
Comment (goldmedal) [inline accio-main/src/main/java/io/accio/main/web/DuckDBResource.java:41]: If it's read-only API, consider to use `@GET`.

File: accio-tests/src/test/java/io/accio/testing/TestDuckDBResource.java [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+
+    @Test
+    public void testSessionSQLWithFailedSQL()
+    {
+        assertWebApplicationException(() -> setDuckDBSessionSQL(FAILED_SQL))
+                .hasErrorMessageMatches(".*Parser Error: syntax error at or near \"xxx\"");
+    }
+
+    @Test
+    public void testSessionSQL()
```
Comment (goldmedal) [inline accio-tests/src/test/java/io/accio/testing/TestDuckDBResource.java:128]: We should have a test to assert if the session SQL is applied to every exist connection in the pool.

File: accio-main/src/main/java/io/accio/main/web/DuckDBResource.java [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    @Path("/query")
+    @Produces(APPLICATION_JSON)
+    public void query(
```
[the remark this answers is not in the document]
Comment (grieve54706) [reply accio-main/src/main/java/io/accio/main/web/DuckDBResource.java:44]: We don't limit the SQL statement type. If we want to limit the API read-only, we need to parse statement to know what type it is.

File: accio-main/src/main/java/io/accio/main/web/DuckDBResource.java (hunk 3 above)
Comment (grieve54706) [reply accio-main/src/main/java/io/accio/main/web/DuckDBResource.java:99]: It will not throw any exception. It still needs handler?

File: accio-base/src/main/java/io/accio/base/client/duckdb/DuckDBConnectorConfig.java [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+    public Path getInitSQLPath()
+    {
+        return settingDir.resolve("init.sql");
+    }
+
+    public Path getSessionSQLPath()
+    {
+        return settingDir.resolve("session.sql");
+    }
```
[the remark this answers is not in the document]
Comment (goldmedal) [reply accio-base/src/main/java/io/accio/base/client/duckdb/DuckDBConnectorConfig.java:32]: When the data source type is DuckDB, we should load them to init the DuckDB.

File: accio-main/src/main/java/io/accio/main/web/DuckDBResource.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (goldmedal) [reply accio-main/src/main/java/io/accio/main/web/DuckDBResource.java:44]: I think we don't need to check the type. The API name just tell user that it's a `read-only` API. If the user submit a DDL or DML, he should expect to get an error.

File: accio-main/src/main/java/io/accio/main/web/DuckDBResource.java (hunk 3 above)
Comment (goldmedal) [reply accio-main/src/main/java/io/accio/main/web/DuckDBResource.java:99]: I prefer to use the same way to do the error handling. It can avoid to miss handle this if we change the behavior of `getSessionSQL` in the future.

File: accio-base/src/main/java/io/accio/base/client/duckdb/DuckDBDataSource.java [hunk 13]
```diff
@@ -66,17 +70,25 @@ public Connection getConnection()
...
-        statement.execute(format("SET s3_endpoint='%s'", duckdbS3StyleStorageConfig.getEndpoint()));
-        statement.execute(format("SET s3_url_style='%s'", duckdbS3StyleStorageConfig.getUrlStyle()));
+        if (cacheStorageConfig instanceof DuckdbS3StyleStorageConfig) {
+            DuckdbS3StyleStorageConfig duckdbS3StyleStorageConfig = (DuckdbS3StyleStorageConfig) cacheStorageConfig;
+            statement.execute(format("SET s3_endpoint='%s'", duckdbS3StyleStorageConfig.getEndpoint()));
+            statement.execute(format("SET s3_url_style='%s'", duckdbS3StyleStorageConfig.getUrlStyle()));
+        }
+        if (duckDBSettingSQL.getSessionSQL() != null) {
+            statement.execute(duckDBSettingSQL.getSessionSQL());
+        }
```
[the remark this answers is not in the document]
Comment (grieve54706) [reply accio-base/src/main/java/io/accio/base/client/duckdb/DuckDBDataSource.java:80]: We can not expect the DuckDB is for what base on SQL is null. Because the SQL may be null at the begin for data source. If we want a basis, we can make DuckDBSettingSQL be null.

[the remark this answers is not in the document]
Comment (goldmedal) [reply accio-base/src/main/java/io/accio/base/client/duckdb/DuckDBDataSource.java:80]: Sure. If `DuckDBSettingSQL` is null, it's a DataSource. Otherwise, it's a PgMetastore.

**Comment by goldmedal** [comment]
> Could you add a test `TestDeployDuckDBRuntime` to test the data source can be switch to DuckDB and config can be changed after server started? I'll keep going to review tomorrow. Thanks

We can implement dynamic deploying DuckDB DataSource in next PR.  Let's only finish the static one here.

File: accio-main/src/main/java/io/accio/main/web/dto/DuckDBQueryDto.java (hunk 6 above)
Comment (grieve54706) [reply accio-main/src/main/java/io/accio/main/web/dto/DuckDBQueryDto.java:10]: Using `PreviewOutputDto` for API of DuckDB may be not appropriate. I don't have good idea for the new name of `PreviewOutputDto` for two API. Make a empty class `DuckDBQueryDto` extend `PreviewOutputDto` may not a good idea too. Could you help me give me a new name?

Comment (goldmedal) [reply accio-main/src/main/java/io/accio/main/web/dto/DuckDBQueryDto.java:10]: Probability, just create a new one called `QueryRresultDto`? I think the format of `PreviewOutputDto` and `DucKDBQueryDto` are same. Both of them have `columns` and `data` (or called `rows`). Use one common DTO is enough.

File: accio-main/src/main/java/io/accio/main/web/DuckDBResource.java [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+        asyncResponse.resume(Response.ok().build());
+    }
+
+    @PATCH
+    @Path("/settings/session-sql")
+    public void appendToSessionSQL(
+            String sql,
+            @Suspended AsyncResponse asyncResponse)
+    {
+        metadata.directDDL(sql);
```
[the remark this answers is not in the document]
Comment (grieve54706) [reply accio-main/src/main/java/io/accio/main/web/DuckDBResource.java:122]: I will close all active connections for it.

File: accio-main/src/main/java/io/accio/main/wireprotocol/WireProtocolSession.java [hunk 15]
```diff
@@ -279,16 +280,10 @@ private void parseMetastoreSemiQuery(String statementName, String statement, Lis
...
         String statementPreRewritten = PostgreSqlRewriteUtil.rewrite(statementTrimmed);
-        SessionContext sessionContext = SessionContext.builder()
-                .setCatalog(getDefaultDatabase())
-                .setSchema(getDefaultSchema())
-                .setEnableDynamic(accioConfig.getEnableDynamicFields())
-                .build();
         try {
             Statement metadataQueryStatement = MetastoreSqlRewrite.rewrite(regObjectFactory,
                     sqlParser.createStatement(statementPreRewritten, PARSE_AS_DECIMAL));
-            String converted = pgMetastore.getSqlConverter().convert(SqlFormatter.formatSql(metadataQueryStatement), sessionContext);
```
Comment (goldmedal) [inline accio-main/src/main/java/io/accio/main/wireprotocol/WireProtocolSession.java:290]: Why don't we convert SQL here?

Comment (grieve54706) [reply accio-main/src/main/java/io/accio/main/wireprotocol/WireProtocolSession.java:290]: In the original code base, the sql coverter from pg metastore is not with any rewriter. And remove this method is helpful for Spliting PgMetastore.

File: accio-main/src/main/java/io/accio/main/pgcatalog/builder/DuckDBFunctionBuilder.java [hunk 16]
```diff
@@ -29,12 +27,11 @@ public class DuckDBFunctionBuilder
         implements PgFunctionBuilder, PgMetastoreFunctionBuilder
 {
     private static final Logger LOG = Logger.get(DuckDBFunctionBuilder.class);
-    private final DuckDBMetadata duckDBMetadata;
+    private final DuckdbClient duckdbClient;
```
[the remark this answers is not in the document]
Comment (grieve54706) [reply accio-main/src/main/java/io/accio/main/pgcatalog/builder/DuckDBFunctionBuilder.java:30]: Actually, it should still keep using `DuckDBMetadata`. But I splited the PgMetastore and Metadata, I can not use PgMetastore when the function builder for DuckDBMetadata.

File: accio-main/src/main/java/io/accio/main/web/DuckDBResource.java [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+    @Path("/settings/session-sql")
+    public void appendToSessionSQL(
+            String sql,
+            @Suspended AsyncResponse asyncResponse)
+    {
+        runAsync(() -> {
+            metadata.getClient().closeAndInitPool();
+            metadata.setSessionSQL(metadata.getSessionSQL() + "\n" + sql);
+            FileUtil.appendToFile(metadata.getSessionSQLPath(), sql);
+        }).whenComplete(bindAsyncResponse(asyncResponse));
```
Comment (goldmedal) [inline accio-main/src/main/java/io/accio/main/web/DuckDBResource.java:150]: This API won't work. Although close and init the pool, we only append and set session SQL but we don't reload the config.

File: accio-main/src/main/java/io/accio/main/connector/duckdb/DuckDBMetadata.java [hunk 18]
```diff
@@ -14,47 +14,69 @@
...
     @Inject
-    public DuckDBMetadata(DuckdbClient duckdbClient)
+    public DuckDBMetadata(
+            ConfigManager configManager)
     {
-        this.duckdbClient = requireNonNull(duckdbClient, "duckdbClient is null");
-        this.pgFunctionBuilder = new DuckDBFunctionBuilder(this);
+        this.configManager = requireNonNull(configManager, "configManager is null");
+        initDuckDBSettingSQLIfNeed();
+        this.duckdbClient = buildDuckDBClient();
```
Comment (goldmedal) [inline accio-main/src/main/java/io/accio/main/connector/duckdb/DuckDBMetadata.java:77]: We should handle executing Init-SQL or Session-SQL failed when the server startup. We should skip the failed SQL to start the server. Then, users can update the SQL by API.

File: accio-main/src/main/java/io/accio/main/pgcatalog/builder/DuckDBFunctionBuilder.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (grieve54706) [reply accio-main/src/main/java/io/accio/main/pgcatalog/builder/DuckDBFunctionBuilder.java:30]: I will add new interface for them.

File: accio-main/src/main/java/io/accio/main/connector/duckdb/DuckDBMetadata.java [hunk 19]
```diff
@@ -155,41 +178,140 @@ public String getPgCatalogName()
...
-            return format("%s[]", handlePgType(type.substring(1)));
-        }
-        else if (!DuckdbTypes.getDuckDBTypeNames().contains(type)) {
-            return "VARCHAR";
-        }
-        return type;
+        // Create client success first to ensure the sql is legal, then close the old one
+        DuckdbClient duckdbClient = buildDuckDBClient();
+        close();
+        this.duckdbClient = duckdbClient;
```
Comment (goldmedal) [inline accio-main/src/main/java/io/accio/main/connector/duckdb/DuckDBMetadata.java:186]: It's acceptable for now. However, there're a potential issue about if the init SQL load large data to DuckDB, it could cost 200% memory system needed when user try to reload client.

File: accio-main/src/main/java/io/accio/main/wireprotocol/PgMetastoreImpl.java [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+                .build();
+    }
+
+    private CacheStorageConfig getCacheStorageConfigIfExists()
+    {
+        try {
+            return configManager.getConfig(CacheStorageConfig.class);
+        }
+        catch (Exception e) {
+            return null;
```
Comment (goldmedal) [inline accio-main/src/main/java/io/accio/main/wireprotocol/PgMetastoreImpl.java:152]:
```suggestion
        catch (Exception e) {
            LOG.warn(e, "%s connector does not support cache storage. Cache is disable.", configManager.getConfig(AccioConfig.class).getDataSourceType().name());
            return null;
```
Let's add some log message.

File: accio-base/src/main/java/io/accio/base/wireprotocol/PgMetastore.java [hunk 21]
```diff
@@ -12,28 +12,32 @@
...
-import io.accio.base.Column;
 import io.accio.base.ConnectorRecordIterator;
 import io.accio.base.Parameter;
+import io.accio.base.client.Client;
 import io.accio.base.sql.SqlConverter;
 
 import java.util.List;
 
-public interface PgMetastore
+public abstract class PgMetastore
```
Comment (goldmedal) [inline accio-base/src/main/java/io/accio/base/wireprotocol/PgMetastore.java:24]: Look like PgMetastoreImpl doesn't need to implement multiple interface now. Can we revert this change?

File: accio-main/src/main/java/io/accio/main/connector/duckdb/DuckDBMetadata.java (hunk 19 above)
Comment (grieve54706) [reply accio-main/src/main/java/io/accio/main/connector/duckdb/DuckDBMetadata.java:186]: You are right, but there is no other way to make sure that the SQL is running correctly.

Comment (goldmedal) [reply accio-main/src/main/java/io/accio/main/connector/duckdb/DuckDBMetadata.java:186]: I prefer to close the original one firstly. The reload only is triggered when `put` initSql, `put` session sql and some reload-required configs been changed. It means user want to discard the original one or expects the DB will be reload.  I don't think we need to keep the old one to be the backup.

File: accio-base/src/main/java/io/accio/base/client/duckdb/DuckdbClient.java [hunk 22]
```diff
@@ -225,4 +250,51 @@ public void close()
             throw new RuntimeException(e);
         }
     }
+
+    public void closeAndInitPool()
```
Comment (goldmedal) [inline accio-base/src/main/java/io/accio/base/client/duckdb/DuckdbClient.java:254]:
```suggestion
    public synchronized void closeAndInitPool()
```
Let's `synchronized` this method to be thread safe.

File: accio-main/src/main/java/io/accio/main/web/DuckDBResource.java [hunk 23]
```diff
@@ -0,0 +1,167 @@
...
+            @Suspended AsyncResponse asyncResponse)
+    {
+        runAsync(() -> {
+            String sessionSQL = metadata.getSessionSQL();
+            metadata.setSessionSQL(sessionSQL + "\n" + sql);
+            try {
+                metadata.getClient().closeAndInitPool();
+            }
+            catch (Exception e) {
+                metadata.setSessionSQL(sessionSQL);
```
Comment (goldmedal) [inline accio-main/src/main/java/io/accio/main/web/DuckDBResource.java:161]:
```suggestion
                metadata.setSessionSQL(sessionSQL);
                metadata.getClient().closeAndInitPool();
```
We should reopen the pool after rollback the session SQL.

**Review by goldmedal — APPROVED** [review]
LGTM. Many thanks.
